frontend/components/atoms/buttons/button.php

```diff
@@ -1,35 +1,21 @@
 <?php
 
-$class  = $args['class'] ?? '';
```

Comment:
I think it's clearer if you define the components data contract at the beginning.
The Dev will visualize immediately the data needed.
Also IDE like PHPstorm will know that this variables exist and detect if any is not used

Reply:
Again the idea here was that we can pass through the ACF fields directly, with no need to write code to pass in the data. I guess it could be defined at the beginning of this file, i was just trying to cut down the amount of code we need to write. I'm open to hear other opinions on this.

Reply:
in big components with lot of data received it will be easier to read the data names. In this case the component is small and it doesn't look complicated, but as it is the component of 'example' it should have them

Reply:
Do you think it would work to add something like this to the top of each template? To me this is cleaner than specifying all the variables and provides the same result in terms of encouraging you to specify what args are passed in and their default values.
`$args = lean_parse_args(
	$args,
	[
		'button_style' => 'default',
		'link' => [
			'url' => '',
			'title' => '',
			'target' => '_self',
		],
	]
);`

Reply:
to me is cleaner like it is.
I also prefer the current one as my IDE detects the variables and doesn't give me the warning that the variable doesn't exist.

Reply:
I'm sure you can change your IDE settings if necessary.
What does everyone else think?

Reply:
no, no. This will warn me if I'm not using a variable in the component

Reply:
it should work though, if you use $ags['link']['url'] in the HTML, because $args is defined at the top.

Reply:
yeah, but you loose the "live linter" of variables. For example, if you declare 10 variables at the beginning but you added 9 in the component's code, the IDE will tell you that you missed one.
With "args" you will always get that the variable exist no matter the array key.

Maybe this is something subjective and it can be worked in the Dev's own way of working

Reply:
I think we ought to have a standard because this will affect everyone. Let me ask that some other people add their opinions.

Reply:
Also, I think the styleguide json file should be updated with this new vars

Reply:
I have mixed feelings with this, I understand Fran's concerns about syntactic analysis with IDE's but also sometimes I think is a waste of time to create a variable for that for example: `$class = $args['class'] ?? '';`, I agree we won't have to couple components with ACF but maybe find an elegant and standarized way to pass props, maybe a configuration Object that can be swaped as needed

Reply:
How would the config object work @arzola ?

I do think Fran is right that we should define the contract at the top each file, showing which variables are required and what the default values are. However, I think that can be done using the proposal I gave above with lean_parse_args - I see this having 2 advantages over the alternative:

1. Easier to read (and quicker to write)
1. Allows for multi-depth variables, which could also be required for repeaters, groups, etc as well as links.

I don't think we should have to convert the ACF output structure, I don't see that it really matters what the input structure you define is, so long as it is defined at the top of the file. Writing code like this seems quite wasteful:

```
$link = get_field( 'link' );
$Load::atom(
  'button',
  [
    'url' => $link['url'],
    'title' => $link['title'],
    'target' => $link['target'],
  ]
);
```

... when this can be fully automated if you don't have to restructure the variables.

Reply:
I don't agree that it's easier to read and quicker to write.
A component's data can be used a lot throughout all the code.

for example if you have the variable $title defined at the beginning you can reuse it in all the code and the IDE will autocomplete the variable and it's cleaner to read:

`
$title = $args['title'] ?? '';

if ( $title ) :
?>

    <h1><?php echo esc_html( $title ); ?></h1>

<?php endif; ?>

<?php if ( ! $title ) : ?>
 <p>Other stuff</p>
<?php endif; ?>
`

The way your proposing would be:

`
$args = lean_parse_args(
	$args,
	[
		'title' => '',
	]
);

if ( $args['title'] ) :
?>

    <h1><?php echo esc_html( $args['title'] ); ?></h1>

<?php endif; ?>

<?php if ( ! $args['title'] ) : ?>
 <p>Other stuff</p>
<?php endif; ?>
`

for me it's a waste of time to open brackets and add keys manually. With vars declared, the IDE will autocomplete them

Reply:
Ok, I think you're right. In any case, I'm making a bigger deal out of this than I needed to. I've updated it back to the original format and pushed the change. Are there any other comments on this PR?

Reply:
As Os, I also have mixed feelings with this and understand the benefits of both approaches. However from a purist reusability/modularity POV, the best approach is the original one, using the vars individually instead of an array, as e.g. you could use a component not necessarily filled in by ACF data, but also for data added manually, so it'd be easier to reuse the components if we know exactly all the required arguments.